### ISSUE TITLE
Enable RTTI

### ISSUE THREAD

**Issue by neobrain** [citra-emu/citra#1054]
RTTI (RunTime Type Information) is a C++ language feature used to identify and cast dynamic objects (i.e. instances of classes with virtual members). Of course, this technically incurs /some/ overhead, since information about the "active" type of an object needs to be stored somewhere in memory. This overhead is the reason why many projects (including Citra) outright disable RTTI.

My proposal is to re-enable RTTI for Citra.

This might be somewhat controversial, but let me clarify that I'm not proposing active usage of RTTI all over the place. If need be, we can actually leave the active usage of RTTI prohibited. The proposal is just about enabling RTTI in the compiler flags.

With that settled, reasons for this proposal are:
* there is literally no overhead from using RTTI for non-virtual classes (since type information will always be known at compile-time), which is what virtually all of Citra's performance/memory-critical data structures should be.
* RTTI overhead for virtual classes seems to be neglegible, at least on GCC as per http://stackoverflow.com/questions/579887/how-expensive-is-rtti/4334421#4334421 . I don't know how things look like for MSVC or clang.
* RTTI might be useful in a few edge cases. GUI code comes to mind, albeit Qt's `moc` provides an alternative anyway (effectively reimplementing RTTI without relying on the language feature).
* Enabling RTTI allows us to safely link against the official boost binaries. There seems to be a workaround for linking them against non-RTTI (http://www.boost.org/doc/libs/1_59_0/doc/html/boost_typeindex/mixing_sources_with_rtti_on_and_.html), but it's not guaranteed to work. Linking boost libraries would be beneficial for including some non-header-only libraries (in particular, test, filesystem, and program_options are of interest for Citra).

However, I'm also genuinely curious if there are indeed any actual drawbacks of RTTI which would affect Citra considerably. Hence feel free to correct me if I'm missing something ;)

tl;dr: I suggest re-enabling RTTI because there are a few minor reasons to do so and there seem to be no actual reasons against it. If people agree, I will make the following changes:
* re-enable RTTI in the build settings
* add a line to the coding style saying that usage of RTTI is strongly discouraged.

**Comment by hrydgard** [comment]
In PPSSPP, RTTI is only disabled because early versions of the Android NDK did not support it, and also Symbian, I believe, which you guys don't support, has broken support if any. The overhead is indeed negligible and I would like to eventually enable it. I don't see any reason why you guys wouldn't.

**Comment by neobrain** [comment]
@hrydgard Interesting point. Good to know that recent versions of the ANDK actually support it, which might have presented a problem in the long run.

Since no one shouted at me for making this proposal yet, I guess I'll prepare a PR for this eventually.

**Comment by neobrain** [comment]
Fixed by  #1238.
